SWC-6318: New entity page title bar

nickgros opened the pull request:
**Requires a new SRC release

https://user-images.githubusercontent.com/17580037/210647288-77f83fbf-0e78-4fe6-85fe-1a9d41a7f3d3.mov


- Rename BasicTitleBar to ProjectTitleBar
- Create new class BasicTitleBar that uses new EntityPageTitleBar React component
- Use BasicTitleBar for all entity pages (except project)
- Delete DockerTitleBar, FileTitleBar, TableTitleBar
- Move file download event handlers from FilesTab to EntityActionControllerImpl
- Clear React-Query downloadList cache on DownloadListUpdatedEvent
- Style and layout changes for consistency with design

nickgros committed:
SWC-6318
- Rename BasicTitleBar to ProjectTitleBar
- Create new class BasicTitleBar that uses new EntityPageTitleBar React component
- Use BasicTitleBar for all entity pages (except project)
- Delete DockerTitleBar, FileTitleBar, TableTitleBar
- Move file download event handlers from FilesTab to EntityActionControllerImpl
- Clear React-Query downloadList cache on DownloadListUpdatedEvent
- Style and layout changes for consistency with design

nickgros commented on src/main/java/org/sagebionetworks/web/client/jsinterop/EntityPageTitleBarProps.java:
```diff
@@ -0,0 +1,71 @@
+import org.sagebionetworks.web.client.widget.entity.menu.v3.EntityActionMenuProps;
+
+@JsType(isNative = true, namespace = JsPackage.GLOBAL, name = "Object")
+public class EntityPageTitleBarProps extends ReactComponentProps {
```
Props for the new React component

nickgros commented on src/main/java/org/sagebionetworks/web/client/jsinterop/entity/actionmenu/EntityActionMenuPropsJsInterop.java:
```diff
@@ -7,24 +7,29 @@
 
 @JsType(isNative = true, namespace = JsPackage.GLOBAL, name = "Object")
-public class EntityActionMenuProps extends ReactComponentProps {
+public class EntityActionMenuPropsJsInterop extends ReactComponentProps {
```
Renamed and created POJO `EntityActionMenuProps` to simplify testing.

nickgros commented on src/main/java/org/sagebionetworks/web/client/jsinterop/reactquery/SynapseReactClientEntityQueryKey.java:
```diff
@@ -8,19 +8,19 @@
 
 @JsType(isNative = true, namespace = JsPackage.GLOBAL, name = "Object")
-public class SynapseReactClientQueryKey {
+public class SynapseReactClientEntityQueryKey {
```
Renamed because this query key structure currently only applies to the entity cache

nickgros commented on src/main/java/org/sagebionetworks/web/client/presenter/EntityPresenter.java:
```diff
@@ -320,14 +322,22 @@ public Widget asWidget() {
     globalAppState.refreshPage();
   }
 
+  @EventHandler
+  public void onDownloadListUpdatedUpdatedEvent(
+    DownloadListUpdatedEvent _event
+  ) {
+    List<String> queryKey = Collections.singletonList("downloadList");
+    queryClient.invalidateQueries(queryKey);
+  }
```
Invalidate the `downloadList` cache items when the download list is updated. Fixes an issue where the count and download list won't immediately update after adding an item to the download cart.

nickgros replied:
https://sagebionetworks.jira.com/browse/SWC-6353

nickgros commented on src/main/java/org/sagebionetworks/web/client/widget/entity/EntityMetadata.java:
```diff
@@ -191,10 +193,6 @@ public void onFailure(Throwable t) {
     );
   }
 
-  public void setVisible(boolean visible) {
-    view.setDetailedMetadataVisible(visible);
-  }
```
Unused

nickgros commented on src/main/java/org/sagebionetworks/web/client/widget/entity/controller/EntityActionControllerImpl.java:
```diff
@@ -697,9 +708,58 @@ private FluentFuture configureFileDownload() {
       actionMenu.setActionVisible(Action.DOWNLOAD_FILE, true);
+
       actionMenu.setActionVisible(Action.ADD_TO_DOWNLOAD_CART, true);
+      actionMenu.setActionListener(
```
This listener was moved from the (now removed) FileTitleBar widget.

nickgros commented on src/main/java/org/sagebionetworks/web/client/widget/entity/controller/EntityActionControllerImpl.java:
```diff
@@ -697,9 +708,58 @@ private FluentFuture configureFileDownload() {
+      );
 
       actionMenu.setActionVisible(Action.SHOW_PROGRAMMATIC_OPTIONS, true);
+      actionMenu.setActionListener(
```
This listener was moved from the (now removed) FileTitleBar widget.

nickgros commented on src/main/java/org/sagebionetworks/web/client/widget/entity/file/BasicTitleBar.java:
```diff
@@ -1,66 +1,78 @@
+  private EntityPageTitleBarProps props;
 
   @Inject
   public BasicTitleBar(
```
This widget now basically just gathers the props for the new React component.

nickgros commented on src/main/java/org/sagebionetworks/web/client/widget/entity/file/BasicTitleBarView.java:
```diff
@@ -1,18 +1,9 @@
 import org.sagebionetworks.web.client.SynapseView;
+import org.sagebionetworks.web.client.jsinterop.EntityPageTitleBarProps;
 
 public interface BasicTitleBarView extends IsWidget, SynapseView {
```
The view becomes a simple wrapper over the React component.

nickgros commented on src/main/java/org/sagebionetworks/web/client/widget/entity/file/DockerTitleBar.java:
```diff
@@ -1,44 +0,0 @@
-package org.sagebionetworks.web.client.widget.entity.file;
```
Removed in favor of BasicTitleBar

nickgros commented on src/main/java/org/sagebionetworks/web/client/widget/entity/file/FileTitleBar.java:
```diff
@@ -1,284 +0,0 @@
-package org.sagebionetworks.web.client.widget.entity.file;
```
Removed in favor of BasicTitleBar

nickgros commented on src/main/java/org/sagebionetworks/web/client/widget/entity/file/ProjectTitleBar.java:
```diff
@@ -0,0 +1,66 @@
+import org.sagebionetworks.web.client.widget.SynapseWidgetPresenter;
+import org.sagebionetworks.web.client.widget.entity.FavoriteWidget;
+
+public class ProjectTitleBar implements SynapseWidgetPresenter {
```
The implementation of `ProjectTitleBar` is the previous implementation of `BasicTitleBar`, with no logical changes--just a rename.

nickgros commented on src/main/java/org/sagebionetworks/web/client/widget/entity/file/ProjectTitleBarView.java:
```diff
@@ -0,0 +1,18 @@
+import org.sagebionetworks.repo.model.EntityType;
+import org.sagebionetworks.web.client.SynapseView;
+
+public interface ProjectTitleBarView extends IsWidget, SynapseView {
```
`ProjectTitleBarView` is the same as the previous implementation of `BasicTitleBarView`--just a rename.

nickgros commented on src/main/java/org/sagebionetworks/web/client/widget/entity/file/ProjectTitleBarViewImpl.java:
```diff
@@ -0,0 +1,95 @@
+import org.sagebionetworks.web.client.DisplayUtils;
+import org.sagebionetworks.web.client.widget.EntityTypeIcon;
+
+public class ProjectTitleBarViewImpl implements ProjectTitleBarView {
```
The implementation of `ProjectTitleBarViewImpl` is the previous implementation of `BasicTitleBarViewImpl`, with no logical changes--just a rename.

nickgros commented on src/main/java/org/sagebionetworks/web/client/widget/entity/file/TableTitleBar.java:
```diff
@@ -1,99 +0,0 @@
-package org.sagebionetworks.web.client.widget.entity.file;
```
Removed in favor of BasicTitleBar

nickgros commented on src/main/java/org/sagebionetworks/web/client/widget/entity/menu/v3/EntityActionMenuImpl.java:
```diff
@@ -135,14 +137,25 @@ public void hideAllActions() {
   }
 
+  @Override
+  public void addPropUpdateListener(Consumer<EntityActionMenuProps> listener) {
```
Every time the action menu props change, we need to re-render the EntityTitleBar and pass along these props, so we'll add a way to register listeners here.

jay-hodgson replied:
Any possibility of stale listeners?  I guess the title bar widget is shared across entities, so this is probably not an issue.

nickgros replied:
Great catch, I took a look and this implementation causes listeners that reference the same BasicTitleBar widget to accumulate. Jumping between entities in the same tab adds one listener each time you visit a new entity page.

Since there only one instance of the BasicTitleBar widget on the page at a time, and that is the only widget that listens to the action menu, I just changed this to only store one listener in state at a time. 

It's not perfect; there might still be memory leak issues, but it's less likely. I'm placing a bet that we will have another chance to re-architect or simplify this before we would need to track multiple listeners.

nickgros commented on src/main/java/org/sagebionetworks/web/client/widget/entity/menu/v3/EntityActionMenuProps.java:
```diff
@@ -0,0 +1,53 @@
+import org.sagebionetworks.web.client.jsinterop.entity.actionmenu.EntityActionMenuLayout;
+import org.sagebionetworks.web.client.jsinterop.entity.actionmenu.EntityActionMenuPropsJsInterop;
+
+public class EntityActionMenuProps {
```
POJO that just wraps the EntityActionMenuPropsJsInterop class, but is easier to test; we don't have to deal with restrictions that come with using JsInterop without a browser.

nickgros commented on src/main/java/org/sagebionetworks/web/client/widget/entity/menu/v3/EntityActionMenuView.java:
```diff
@@ -1,28 +1,12 @@
-import org.sagebionetworks.web.client.jsinterop.entity.actionmenu.ActionConfiguration;
-import org.sagebionetworks.web.client.jsinterop.entity.actionmenu.EntityActionMenuDropdownMap;
-import org.sagebionetworks.web.client.jsinterop.entity.actionmenu.EntityActionMenuLayout;
+import org.sagebionetworks.web.client.jsinterop.entity.actionmenu.EntityActionMenuPropsJsInterop;
```
```suggestion
```

nickgros committed:
Update src/main/java/org/sagebionetworks/web/client/widget/entity/menu/v3/EntityActionMenuView.java

nickgros commented on src/main/java/org/sagebionetworks/web/client/widget/entity/menu/v3/EntityActionMenuView.java:
```diff
@@ -1,28 +1,11 @@
-import org.sagebionetworks.web.client.jsinterop.entity.actionmenu.EntityActionMenuLayout;
 
 public interface EntityActionMenuView extends IsWidget {
-  void setPresenter(Presenter presenter);
```
Presenter is no longer needed

nickgros commented on src/main/java/org/sagebionetworks/web/client/widget/entity/menu/v3/EntityActionMenuView.java:
```diff
@@ -1,28 +1,11 @@
 public interface EntityActionMenuView extends IsWidget {
-  void setPresenter(Presenter presenter);
-
-  void configure(
-    Map<Action, ActionConfiguration> actionConfiguration,
-    EntityActionMenuDropdownMap menuConfiguration,
-    EntityActionMenuLayout layout
-  );
-
-  void setLayout(EntityActionMenuLayout layout);
```
Replaced both with `configure(EntityActionMenuProps)`

nickgros commented on src/main/java/org/sagebionetworks/web/client/widget/entity/menu/v3/EntityActionMenuViewImpl.java:
```diff
@@ -95,21 +67,6 @@ private void renderLoaderComponent() {
     loaderComponent.render(node);
   }
 
-  private EntityActionMenuProps getProps() {
```
`getProps` was moved to the EntityActionMenuProps Java object (now called `EntityActionMenuProps.toJsInterop`)

nickgros commented on src/main/java/org/sagebionetworks/web/client/widget/table/explore/TableEntityWidgetV2.java:
```diff
@@ -335,7 +335,10 @@ private void configureActions() {
         );
     }
-    this.actionMenu.setActionVisible(Action.EDIT_TABLE_DATA, true);
+    this.actionMenu.setActionVisible(
```
Fix bug where the EDIT_TABLE_DATA action should not be visible on non-editable table types (datasets, materialized views)

nickgros replied:
https://sagebionetworks.jira.com/browse/SWC-6352

nickgros commented on src/main/resources/org/sagebionetworks/web/client/widget/docker/DockerRepoWidgetViewImpl.ui.xml:
```diff
@@ -15,61 +15,65 @@
-          ui:field="dockerModifiedAndCreatedContainer"
-        />
-      </b:Row>
+      <g:FlowPanel addStyleNames="entity-page-side-margins">
```
Fix side margins

nickgros commented on src/main/resources/org/sagebionetworks/web/client/widget/entity/EntityMetadataViewImpl.ui.xml:
```diff
@@ -6,11 +6,14 @@
   xmlns:w="urn:import:org.sagebionetworks.web.client.widget"
 >
-  <g:HTMLPanel>
+  <g:HTMLPanel addStyleNames="entityMetadata entity-page-side-margins">
```
Fix side margins

nickgros commented on src/main/resources/org/sagebionetworks/web/client/widget/entity/EntityMetadataViewImpl.ui.xml:
```diff
@@ -6,11 +6,14 @@
-        <bh:Span addStyleNames="boldText margin-right-5" text="Synapse ID:" />
+        <bh:Span
+          addStyleNames="boldText margin-right-5"
+          text="Project SynID:"
```
This section is now only shown for projects, so we clarify that this is the Project SynID (specified by design)

nickgros commented on src/main/resources/org/sagebionetworks/web/client/widget/entity/restriction/v2/RestrictionWidgetViewImpl.ui.xml:
```diff
@@ -36,7 +36,7 @@
       <w:HelpWidget
         helpMarkdown="Conditions for use describes data use requirements that must be fulfilled before downloading."
-        href="https://help.synapse.org/docs/Conditions-for-Use.2009596938.html"
+        href="https://help.synapse.org/docs/Sharing-Settings,-Permissions,-and-Conditions-for-Use.2024276030.html#SharingSettings,Permissions,andConditionsforUse-ConditionsforUse"
```
Fix broken docs link

jay-hodgson commented on src/main/java/org/sagebionetworks/web/client/presenter/EntityPresenter.java:
```diff
@@ -320,14 +322,22 @@ public Widget asWidget() {
+  public void onDownloadListUpdatedUpdatedEvent(
+    DownloadListUpdatedEvent _event
+  ) {
+    List<String> queryKey = Collections.singletonList("downloadList");
```
Consider a constants file that contains all of the react query keys

nickgros committed:
SWC-6318 - Code review changes
- In EntityActionMenu change addPropUpdateListener to setPropUpdateListener, only tracking one listener at a time. Currently, only one instance of one widget, the title bar, needs to be updated. This helps prevent accumulating listeners that reference the same instance of the title bar.
- Use constants for query keys

nickgros committed:
Merge branch 'develop' into SWC-6318

nickgros committed:
SWC-6318
- Picked wrong EntityActionMenu in merge conflict resolution
